Add support for refreshing env vars and fix dev build

ajbozarth opened the pull request:
- Adds support for refreshing the env vars in a node's properties
- Adds (back) automatic detection of env vars on node add
- Updates build scripts and documentation to support development with Elyra

paired with https://github.com/elyra-ai/elyra/pull/1654
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

lresende commented:
> @ajbozarth I know this is just a WIP, but I think we could do a top level button for refreshing the properties instead of just a button for refreshing the string array form control
> 
> that way we could refresh the entire form if we ever have add dynamic properties as a feature
> 
> thoughts?

Where is the button now? Having the button on the top if it's only refreshing vars would be very confusing, but when we have other functionality might be a good option... The other thing is, what should be the user experience? Refresh sections vs refresh all?

bourdakos1 commented:
@lresende with the potential dynamic properties feature, entire sections could not exist and then exist after a refresh, so a per section refresh wouldn't work. Having a high level "refresh all" button might be a little less clear than a button attached to the env vars section, but I think if we do use this partial refresh button approach we will get locked into it. If the "refresh all" approach ends up being too confusing, we can address the problem

I think if we handle this correctly, the user should have to click refresh vary rarely

marthacryan commented on README.md:
````diff
@@ -42,13 +42,53 @@ npm install @elyra/pipeline-editor
+```sh
+cd node_modules/react
+yarn link
+cd ../node_modules/@types/react
````
Should this be `cd ../@types/react`?

ajbozarth replied:
yes

marthacryan commented on README.md:
````diff
@@ -42,13 +42,53 @@ npm install @elyra/pipeline-editor
+yarn link
+```
+
+or if you're running with Elyra you can use make:
````
I think this step is kind of confusing - what exactly is it replacing out of the steps above?

marthacryan commented on README.md:
````diff
@@ -42,13 +42,53 @@ npm install @elyra/pipeline-editor
+If you're connecting to a Jupyterlab extension run:
+
+```sh
+npm run jl-link
````
Might be good to also include the build steps after this. Can you run `make clean install`? And what other steps might you need to install the JL extension with the linked package?

marthacryan replied:
Also - maybe unlink steps as well?

marthacryan replied:
Also watch?

ajbozarth commented:
@bourdakos1 @marthacryan I've pushed the build updates to both PRs. Feel free to double check everything works. I'll be returning to working on the actual functionality now.

ajbozarth commented:
I just pushed some updates to the PRs with:

- Fixed bug where new vars weren't saved on close
- Added (back) support for env var detection on node add
- Updated callback function to be more generic so it can be extended with more properties in the future

The last point was to partially address @bourdakos1 and @lresende discussion above. I believe for now leaving the refresh button inline where it actual applies makes sense, but I updated the code so that when we make it more generic in the future we only have the move the button, the actual callback function is already ready to support a full refresh.

With these changes I'm also moving this out of WIP for review

bourdakos1 commented:
@ajbozarth You should be able to change `peerDependencies` to `true` in eslint to get rid of all the react lint warnings: https://github.com/ajbozarth/pipeline-editor/blob/f64a4918346149ae5112e128c0cfa4c2db95831a/.eslintrc.js#L62

bourdakos1 commented on packages/pipeline-editor/src/CustomFormControls/StringArrayControl.tsx:
```diff
@@ -316,6 +316,17 @@ function StringArrayControl({ placeholder, format }: Props) {
 
+  const handleNewValues = useCallback(async () => {
+    const updatedProperties = await actionHandler?.(
+      "update_properties",
```
can we change this to `request_properties_update` or `refresh_properties`? I keep thinking `update_properties` is the builtin `UPDATE_PROPERTY` action

ajbozarth replied:
done, will be in next commit

bourdakos1 commented on packages/pipeline-editor/src/CustomFormControls/StringArrayControl.tsx:
```diff
@@ -388,6 +399,16 @@ function StringArrayControl({ placeholder, format }: Props) {
               Browse
             </button>
           )}
+          {format === "properties" && (
```
yea, let's change this to a `canXXX` boolean like mentioned above. The `format` prop is more intended for potential regex of common formats like date, time, email, etc...

ajbozarth replied:
done, will be in next commit

bourdakos1 commented on packages/pipeline-editor/src/CustomFormControls/StringArrayControl.tsx:
```diff
@@ -388,6 +399,16 @@ function StringArrayControl({ placeholder, format }: Props) {
+                handleNewValues();
+              }}
+            >
+              Detect Updates
```
can we change this to `Refresh`? `Detect Updates` gives me the impression that it will keep automatically detecting updates

ajbozarth replied:
done, will be in next commit

bourdakos1 commented on packages/pipeline-editor/src/PipelineController/index.ts:
```diff
@@ -133,6 +133,10 @@ class PipelineController extends CanvasController {
     data.nodeTemplate.app_data.filename = item.path;
+    const properties = await item.onPropertiesUpdateRequested?.({
+      filename: item.path,
+    });
```
With the coming changes where we fetch the palette from the server, we will need to request properties whenever the properties panel opens, so I think we should move this there as well maybe?

ajbozarth replied:
I'm not sure where you think we should do that, but it sound like it could just be added once we make that change

bourdakos1 replied:
okay, Im fine with that

ajbozarth commented on packages/pipeline-editor/src/CustomFormControls/StringArrayControl.tsx:
```diff
@@ -26,7 +26,7 @@ import { StringArrayValidatorOptions } from "./validators";
 interface Props extends StringArrayValidatorOptions {
   placeholder?: string;
-  format?: "file";
+  format?: string;
```
done, will be in next commit

bourdakos1 commented on packages/pipeline-editor/src/PropertiesPanel/index.tsx:
```diff
@@ -105,6 +107,9 @@ function PropertiesPanel({
           switch (id) {
             case "browse_file":
               return await onFileRequested?.(data);
+            case "refresh_properties":
+              data.filename = selectedNode.app_data.filename;
+              return await onPropertiesUpdateRequested?.(data);
```
I missed this earlier, but this makes me uncomfortable. 
can we do something like this instead:
```ts
return await onPropertiesUpdateRequested?.({
    ...data,
    filename: selectedNode.app_data.filename
});
```

but also, do we need to send back all the properties? Can we just do `onPropertiesUpdateRequested?.(nodeID);` instead?

ajbozarth replied:
I can make that change

> do we need to send back all the properties

we're not sending back all the properties, for this action `data` is `{ env_vars: items }` as set in `handleRefreshProperties` in `StringArrayControl.tsx`.

bourdakos1 replied:
oh gotcha, do we need to send the `env_vars` data back?

ajbozarth replied:
I guess we dont need to if I move the merging code to the editor side, I can give that a try

ajbozarth replied:
update pushed

bourdakos1 replied:
Oh sorry, I misread that. The merging logic should stay on the jupyter side, but the jupyter side should also already have the most up-to-date data, right?

lresende commented on package.json:
```diff
@@ -57,18 +57,22 @@
+    "lint": "eslint . --ignore-path .gitignore --ext .ts,.tsx --max-warnings=0",
+    "format": "prettier --ignore-path .gitignore --write \"**/*.{tsx,ts,js,css,html,json}\"",
+    "format:check": "prettier --ignore-path .gitignore --check \"**/*.{tsx,ts,js,css,html,json}\"",
+    "lint-fix": "eslint . --fix --ignore-path .gitignore --ext .ts,.tsx"
```
can we have the two lint commands together after the format one.

bourdakos1 replied:
oh yea, and could we also do `lint:fix`

ajbozarth replied:
fixed in next commit

lresende commented:
When I am running on a clean env, `yarn clean` fails during unlink:
```
yarn clean install build link-all
yarn run v1.22.4
$ lerna run clean; rm -rf node_modules install build link-all
lerna notice cli v4.0.0
lerna info Executing command in 2 packages: "yarn run clean"
lerna ERR! yarn run clean exited 1 in '@elyra/pipeline-services'
lerna ERR! yarn run clean stdout:
$ rm -rf node_modules; rm -rf dist; yarn unlink
info Visit https://yarnpkg.com/en/docs/cli/unlink for documentation about this command.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! yarn run clean stderr:
error No registered package found called "@elyra/pipeline-services".
error Command failed with exit code 1.
lerna ERR! yarn run clean exited 1 in '@elyra/pipeline-services'
✨  Done in 1.44s.
```

Maybe we should tolerate (fail silently) when there is nothing to onlink...

lresende commented:
When running the commands separately, I am seeing some warnings. Can we resolve these?
```
yarn build
yarn run v1.22.4
$ lerna run build
lerna notice cli v3.22.1
lerna info Executing command in 2 packages: "yarn run build"
lerna info run Ran npm script 'build' in '@elyra/pipeline-services' in 2.3s:
$ tsc
lerna info run Ran npm script 'build' in '@elyra/pipeline-editor' in 14.8s:
$ microbundle --css inline -f cjs
rpt2: options error TS5089: Option 'jsxFactory' cannot be specified when option 'jsx' is 'react-jsx'.
Build "pipelineEditor" to dist:
      17.2 kB: index.js.gz
      15.2 kB: index.js.br
lerna success run Ran npm script 'build' in 2 packages in 17.1s:
lerna success - @elyra/pipeline-editor
lerna success - @elyra/pipeline-services
✨  Done in 17.75s.
```

lresende commented:
Also, could we update the `Makefile` with:
- make clean --> yarn clean
- make build or install --> yarn install && yarn build
- make link-dev --> link-all 

The last target name could be adjusted, but we should be consistent and use the same from elyra repo, no?

bourdakos1 commented:
> When running the commands separately, I am seeing some warnings. Can we resolve these?
> 
> ```
> yarn build
> yarn run v1.22.4
> $ lerna run build
> lerna notice cli v3.22.1
> lerna info Executing command in 2 packages: "yarn run build"
> lerna info run Ran npm script 'build' in '@elyra/pipeline-services' in 2.3s:
> $ tsc
> lerna info run Ran npm script 'build' in '@elyra/pipeline-editor' in 14.8s:
> $ microbundle --css inline -f cjs
> rpt2: options error TS5089: Option 'jsxFactory' cannot be specified when option 'jsx' is 'react-jsx'.
> Build "pipelineEditor" to dist:
>       17.2 kB: index.js.gz
>       15.2 kB: index.js.br
> lerna success run Ran npm script 'build' in 2 packages in 17.1s:
> lerna success - @elyra/pipeline-editor
> lerna success - @elyra/pipeline-services
> ✨  Done in 17.75s.
> ```

I couldn't figure out how to get rid of the `jsxFactory` warning, I think it was a bug with microbundle at the time because this was right when react 17 came out. I'll look to see if it has been resolved

bourdakos1 commented:
> Also, could we update the `Makefile` with:
> 
> * make clean --> yarn clean
> * make build or install --> yarn install && yarn build
> * make link-dev --> link-all
> 
> The last target name could be adjusted, but we should be consistent and use the same from elyra repo, no?

I use this in `vscode-pipeline-editor`, it create a make target for all npm scripts:
```makefile
.PHONY: help

help:
# http://marmelab.com/blog/2016/02/29/auto-documented-makefile.html
	@grep -E '^[a-zA-Z0-9_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'

	@echo "\nNPM Scripts:"
	@echo "$(shell node -e 'for (const s in require("./package.json").scripts) {console.log(s.replace(/:/g, "-"));}')" | tr ' ' '\n' | awk '{printf "- \033[36m%-30s\033[0m\n", $$1}'

# addopted from https://github.com/kvz/fakefile
define npm_script_targets
TARGETS := $(shell node -e 'for (const s in require("./package.json").scripts) {console.log(s.replace(/:/g, "-"));}')
$$(TARGETS):
	yarn $(subst -,:,$(MAKECMDGOALS))

.PHONY: $$(TARGETS)
endef

$(eval $(call npm_script_targets))
```

ajbozarth commented:
update pushed, includes:

- fix for yarn clean to soft fail on unlink when run before linking
- renamed lint-fix to lint:fix
- updated Makefile with streamlined build steps, note that `make dev-link` can be run anytime after `make clean` and only symlinks the directories, whether or not the code is built.
- updated README to match streamlined make steps
